### PR TITLE
refactor tests to get rid of duplication

### PR DESCRIPTION
**Pull request by rongou**
@RAMitchell @hcho3

### CODE REVIEW

File: tests/cpp/helpers.cc [hunk 1]
```diff
@@ -142,4 +143,38 @@ std::shared_ptr<xgboost::DMatrix>* CreateDMatrix(int rows, int columns,
   return static_cast<std::shared_ptr<xgboost::DMatrix> *>(handle);
 }
 
+std::unique_ptr<DMatrix> CreateSparsePageDMatrix() {
+  // Create sufficiently large data to make two row pages
+  dmlc::TemporaryDirectory tempdir;
+  const std::string tmp_file = tempdir.path + "/big.libsvm";
+  CreateBigTestData(tmp_file, 12);
+  DMatrix * dmat = DMatrix::Load(
+      tmp_file + "#" + tmp_file + ".cache", true, false, "auto", 64UL);
```
Comment (hcho3) [inline tests/cpp/helpers.cc]: Let's wrap pointer in the smart pointer immediately, for the sake of exception safety.
```cpp
std::unique_ptr<DMatrix> dmat = std::unique_ptr<DMatrix>(DMatrix::Load(
      tmp_file + "#" + tmp_file + ".cache", true, false, "auto", 64UL));
```

File: tests/cpp/helpers.cc [hunk 2]
```diff
@@ -142,4 +143,38 @@ std::shared_ptr<xgboost::DMatrix>* CreateDMatrix(int rows, int columns,
...
+  int64_t batch_count = 0;
+  int64_t row_count = 0;
+  for (const auto &batch : dmat->GetRowBatches()) {
+    batch_count++;
+    row_count += batch.Size();
+  }
+  EXPECT_EQ(batch_count, 2);
+  EXPECT_EQ(row_count, dmat->Info().num_row_);
+
+  return std::unique_ptr<DMatrix>(dmat);
```
Comment (hcho3) [inline tests/cpp/helpers.cc]:
```suggestion
  return dmat;
```

File: tests/cpp/helpers.cc (hunk 1 above)
Comment (rongou) [reply tests/cpp/helpers.cc]: Done.

File: tests/cpp/helpers.cc (hunk 2 above)
Comment (rongou) [reply tests/cpp/helpers.cc]: Done.

**Comment by rongou** [comment]
@RAMitchell @hcho3 do you know why the python test is failing? This PR only changes c++ tests, so shouldn't affect python tests at all, right?

**Comment by hcho3** [comment]
@rongou The failure is due to out of memory error.
I am taking a day off tomorrow to sort out CI issues. In particular, all Linux tests should be moved to Jenkins to use machines with bigger memory.

**Pull request merged**